xds/src/main/java/io/grpc/xds/EdsLoadBalancer2.java

```diff
@@ -0,0 +1,421 @@
+        return true;
+      }
+
+      @Override
+      public void onEndpointChanged(EndpointUpdate update) {
+        logger.log(XdsLogLevel.DEBUG,
+            "Received endpoint update from xDS client {0}: {1}", xdsClient, update);
+        if (logger.isLoggable(XdsLogLevel.INFO)) {
+          logger.log(
+              XdsLogLevel.INFO,
+              "Received endpoint update: cluster_name={0}, {1} localities, {2} drop categories",
+              update.getClusterName(), update.getLocalityLbEndpointsMap().size(),
+              update.getDropPolicies().size());
+        }
+        lbHelper.updateDropPolicies(update.getDropPolicies());
+        Map<Locality, LocalityLbEndpoints> localityLbEndpoints = update.getLocalityLbEndpointsMap();
+        endpointAddresses = new ArrayList<>();
+        prioritizedLocalityWeights = new HashMap<>();
+        for (Locality locality : localityLbEndpoints.keySet()) {
+          LocalityLbEndpoints localityLbInfo = localityLbEndpoints.get(locality);
+          int priority = localityLbInfo.getPriority();
+          for (LbEndpoint endpoint : localityLbInfo.getEndpoints()) {
+            if (endpoint.isHealthy()) {
+              EquivalentAddressGroup eag =
+                  AddressFilter.setPathFilter(
+                      endpoint.getAddress(),
+                      Arrays.asList(priorityName(priority), localityName(locality)));
+              endpointAddresses.add(eag);
+            }
+          }
+          if (!prioritizedLocalityWeights.containsKey(priority)) {
+            prioritizedLocalityWeights.put(priority, new HashMap<Locality, Integer>());
+          }
+          prioritizedLocalityWeights.get(priority).put(
+              locality, localityLbInfo.getLocalityWeight());
+        }
+        if (lb == null) {
+          lb = lbRegistry.getProvider(PRIORITY_POLICY_NAME).newLoadBalancer(lbHelper);
+        }
+        if (localityPickingPolicy != null && endpointPickingPolicy != null) {
+          PriorityLbConfig config = generatePriorityLbConfig(cluster, edsServiceName,
+              lrsServerName, localityPickingPolicy, endpointPickingPolicy, lbRegistry,
+              prioritizedLocalityWeights);
+          // TODO(chengyuanzhang): to be deleted after migrating to use XdsClient API.
+          Attributes attributes;
+          if (lrsServerName != null) {
+            attributes =
+                resolvedAddresses.getAttributes().toBuilder()
+                    .set(XdsAttributes.ATTR_CLUSTER_SERVICE_LOAD_STATS_STORE, loadStatsStore)
+                    .build();
+          } else {
+            attributes = resolvedAddresses.getAttributes();
+          }
+          lb.handleResolvedAddresses(
+              resolvedAddresses.toBuilder()
+                  .setAddresses(endpointAddresses)
+                  .setAttributes(attributes)
+                  .setLoadBalancingPolicyConfig(config)
+                  .build());
+        }
```

Comment:
Kind of yes. Lines for putting `LoadStatsStore` into the attribute is temporary and will be deleted after the XdsClient has API to directly access per-locality stats object. Also, there are really only a few statement, but just involving the verbose builder pattern.

However, I also noticed that it would be better to make the latest `ResolvedAddresses` as a field in `ChildLbState`, as each `ChildLbState` instance operates according to a specific address update, so it should be bound with that `ResolvedAddresses`.